README.md
# pds-cdwh
Tools for parsing data from the pds data pull

## pipeline
The pre-processing pipeline looks something like this:

```
cat ${dataroot}${drug}/observation_fact_dx.csv |./procCSV/filter-pids.pl -e ICD9  --nodryRun -i - ;
cat ${dataroot}${drug}/observation_fact_dx.csv |./procCSV/filter-pids.pl -e ICD10 --nodryRun -i - ;
mkdir -p ../out/${drug}/all ;
cat ../out/${drug}/ICD9/bleeding.pid-code ../out/${drug}/ICD10/bleeding.pid-code|sort -u > ../out/${drug}/all/bleeding.pid-code ;
./procCSV/xlate-ae-patvars.pl -e all -a bleeding -d ${drug} --nodryRun  > ../out/${drug}/all/bleeding.patvars 2 ; 
```


CMDUtil/Changes
Revision history for Perl extension CMDUtil.

0.01  Wed Nov 13 23:20:45 2019
	- original version; created by h2xs 1.23 with options
		-XAn CMDUtil



procCSV/filter-pids.pl
#!/usr/bin/perl
use warnings;
use strict;
use 5.010;

use CMDUtil::lib::CMDUtil;

##################
# Set-up
##################

# read-in the default values from config.json
my $settingsFilename = './procCSV/config.json';
# the json has config for multiple programs, just read params for "filter-pids"
our $config = "filter-pids"; 

our %opt;
our $dxFilename;
# now read the config file and get the defaults


CMDUtil::getDefaults("./procCSV/config.json", $config);
%opt=%CMDUtil::opt;


$dxFilename = $opt{'baseDx'}.$opt{'drugType'}.$opt{'dxFilename'};

use Getopt::Long qw< :config auto_version bundling no_ignore_case >;
use Pod::Usage qw(pod2usage);

# get any override values from command line, and/or print documentation
GetOptions (
    "e|encoding=s"   => \$opt{'encoding'},
    "a|aeType=s"     => \$opt{'aeType'},
    "d|drugType=s"   => \$opt{'drugType'},
    "o|outpath=s"    => \$opt{'outPath'},
    "i|inputFile=s"  => \$opt{'dxFilename'},
    "b|baseDx=s"     => \$opt{'baseDx'},
    "f|fileDx=s"     => \$opt{'fileDx'},
    "c|codesPath=s"  => \$opt{'codesPath'},
    "s|settingsFile=s" => \$opt{'settingsFilename'},
    "r|dryRun!"       => \$opt{'dryrun'},
    "v|verbose!"     => \$opt{'verbose'}, 
    
    # Standard options
    "usage"        => sub { pod2usage(2) },
    "help|?"       => sub { pod2usage(1) },
    "manual"       => sub {pod2usage( -exitstatus => 0, -verbose => 2) },
    )
    or pod2usage(2);
pod2usage(1) if $opt{'help'};
pod2usage(-exitval => 0, -verbose => 2) if $opt{'man'};

# re-read configuration if user specified a differieng json file
if(defined $opt{'settingsFilename'} && $opt{'settingsFilename'} ne "") {
    CMDUtil::info("Reading new defaults from:$opt{'settingsFilename'}");
    getDefaults($opt{'settingsFilename'});
} else {
    $opt{'settingsFilename'} = $settingsFilename;
}


# if the input file isn't explicitly specified, construct it from other options
if( ! defined $opt{'dxFilename'} || $opt{'dxFilename'} eq "") {
    $opt{'dxFilename'}="$opt{'baseDx'}$opt{'drugType'}/$opt{'fileDx'}";
    CMDUtil::info("dfFilename is NOT defined, or else it is defined but it's empty, set using baseDx, drugType, fileDx options: $opt{'dxFilename'}");
}

# Finally, dump out the configuration
%CMDUtil::opt = %opt;
CMDUtil::info("--- SETTINGS: ---"); # only prints if verbose is actually on
use Data::Dumper;
CMDUtil::info( Dumper \%opt) ;

# start the dry run
if( $opt{'dryrun'} ) {
    CMDUtil::info("\n******************************* DRY RUN **********************************");
}

##################
# Parse
##################

my $codes="$opt{'codesPath'}/$opt{'aeType'}.$opt{'encoding'}.codes";
open(CODES,"<",$codes) or die("Can't open $codes");
chomp(my @ICDCodes=<CODES>);

my $pidsPath="$opt{'outPath'}/$opt{'drugType'}/$opt{'encoding'}";
if( ! -e $pidsPath) {
    use File::Path qw(make_path);
    make_path($pidsPath) or die("Can't create ${pidsPath}");
} else {
    if( ! -d $pidsPath || ! -W _ ) {	die("Can't write files to $pidsPath"); }
}

my $matchName="${pidsPath}/$opt{'aeType'}.pid-code";
my $notMatchName="${pidsPath}/NOT$opt{'aeType'}.pid-code";

if( -e $matchName)    { CMDUtil::info(" ! $matchName will be overwritten"); } 
else                  { CMDUtil::info("Writing to $matchName");}
if( -e $notMatchName) { CMDUtil::info(" ! $notMatchName will be overwritten"); }
else                  { CMDUtil::info("Writing to $notMatchName"); }

if(! $opt{'dryrun'} ) {
    open(MATCH,">",${matchName}) or die ("Can't open $matchName");
    open(NONMATCH,">",${notMatchName}) or die ("Can't open $notMatchName");
}

# these constants could maybe be in the config:
use constant MAX_LINES => 300000000; # ~ 70GB, for the STDIN case where you can't compute the lines in advance
use constant STEP_SIZE => 1000; # number of lines to read before updating the progress bar
my $fh;
my $count;
if($opt{'dxFilename'} eq "-") {
    $fh = *STDIN;
    $count = MAX_LINES;
    CMDUtil::info("reading from STDIN");
} else {
    open($fh,"<",$opt{'dxFilename'}) or die ("Can't open input file $opt{'dxFilename'}");
    $count = `wc -l < $opt{'dxFilename'}`; die "wc failed: $?" if $?; chomp($count);
    CMDUtil::info("reading from $opt{'dxFilename'}");
}

use Term::ProgressBar;
my $progress = Term::ProgressBar->new($count);
my $lineNum=0;
while( <$fh>) {
    $lineNum++;
    if( ($lineNum % STEP_SIZE) == 0) {
	$progress->update($lineNum);
    }
    if( ! $opt{'dryrun'} ){
	my ($enc, $pid, $code, @scrap) = split(",");
	my $matchFound=0;
	foreach my $i (@ICDCodes) {
	    if($code eq "\"$opt{'encoding'}:".$i."\"") {
		print MATCH "${pid}\t${code}\n";
		$matchFound=1;
	    }
	}
	if($matchFound == 0) {
	    print NONMATCH "${pid}\t${code}\n";
	}
    }
}

$progress->update($count);
CMDUtil::info("-----Done. Next: xlate-ae-patvars.pl -----");



######################### Documentation ######################### 

__END__

=head1 NAME

filter-pids.pl - filters i2b2-sourced 'Observation Fact' diagnosis CSV file into patient numbers with/without the adverse event identified by given ICD codes.

=head1 SYNOPSIS

 ./procCSV/filter-pids.pl [--encoding <ICD10|ICD9>] [--aeType <ae-type>] [--drugType <DOAC|GENT>] [--inputFile <dx-filename>] [--codesPath <codes-path>]  [--fileDx <file-dx> | ([--outPath <out-path>] [--baseDx <base-dx>])]

=head1 DESCRIPTION

B<filter-pids> filters i2b2-sourced 'Observation Fact' diagnosis CSV file into patient numbers and codes that are for adverse and non-adverse events, as identified by given ICD codes. 

 The input i2b2-sourced CSV will be determined by options -b,-d,-f; e.g.:
   Default <base-dx><drug-type><file-dx> = ${dxFilename}
 Data will be output to <ae-type>.pid-code and NOT<ae-type>.pid-code
   under <outPath>/<drugType>/<encoding>
Notably, the 'NOT<ae-type>.pid-code' file will contain the pids for all patients that suffer an adverse event, but that were also diagnosed with a code that signifies a non-adverse event. That is, it's very likely that the 'NOT...' file will contain records for all pids, including those in the adverse event file.
 ex.: 
  ./procCSV/filter-pids.pl -e ICD10 -a bleeding -d DOAC
 reads codes from $opt{'codesPath'}/bleeding.ICD10.codes
 creates files:
  $opt{'outPath'}/DOAC/ICD10/bleeding.pid-code
  $opt{'outPath'}/DOAC/ICD10/NOTbleeding.pid-code

=head1 OPTIONS

=for pod2usage:
  help message for pod2usage

=over

=item B<-e>, B<--encoding>

Discerns which file to use in data.in/codes/<ae-type>.<encoding>.codes
                      [$opt{'encoding'}]    

=item B<-a>, B<--aeType>

Discerns which file to use in data.in/codes/<ae-type>.<encoding>.codes
                      [$opt{'aeType'}] 

=item B<-d>, B<--drugType>

Discerns which drug type to use (e.g., DOAC or GENT)
                      [$opt{'drugType'}]

=item B<-i>, B<--inputFile>

If '-', reads from STDIN. Overrides -b, -f (below)
                      [${dxFilename}]

=item B<-b>, B<--baseDx>

Base path in which to find the i2b2-sourced 'Observation Fact diagnosis CSV file
                      [$opt{'baseDx'}]

=item B<-f>, B<--fileDx>

Filename of the i2b2-sourced 'Observation Fact diagnosis CSV file
                      [$opt{'fileDx'}]

=item B<-c>, B<--codesPath>

Path to ICD9/ICD10 codes for the advers event (aeType)
                      [$opt{'codesPath'}]

=item B<-s>, B<--settingsFile>

Path to a json containing the programs default values for all the arguments described here. This file is found in B<./procCSV/config.json> by default.


=item B<-r>, B<--dryrun>

Don't do any heavy lifting, just open files, create paths - don't read the big files, don't over-write anything, just print the 'informative' messages if verbose is on. Messages are prepended with '[D]' in dryrun mode.

=item B<-v>, B<--verbose>

Enable verbose mode

=item B<--man>

Print man page.

=back

=head1 TODO

Add in support for more different drugs
Add ICD cods for other adverse events
Make i2b2 file format configurable

=head1 AUTHOR

Kimberly Robasky E<lt>krobasky@gmail.com<gt>


=head1 BUGS

None known.

=head1 SEE ALSO

xlate-ae-patvars.pl

=head1 COPYRIGHT and LICENSE

MIT License
Copyright (c) 2019 Kimberly Robasky/RENCI, UNC Chapel Hill


=cut


analyze/characterize.py
# #!/usr/bin/python
# use `conda activate doac` to run this
# just some exploratory code right now
import pandas as pd;
from scipy import stats;
import numpy as np;

def compare_dist(df, cases, controls, census, colname, tagname):

    popTotalCurrent=df[df[colname]==1].describe().loc['count','age'];
    casesTotalCurrent=cases[cases[colname]==1].describe().loc['count','age'];
    controlsTotalCurrent=controls[controls[colname]==1].describe().loc['count','age'];

    pval=popTotalCurrent/popTotalPatients;

    allSig=stats.binom_test(popTotalCurrent, n=popTotalPatients, p=census, alternative='two-sided')
    casesSig=stats.binom_test(casesTotalCurrent, n=AETotalPatients, p=pval, alternative='two-sided');
    controlsSig=stats.binom_test(controlsTotalCurrent, n=AETotalPatients, p=pval, alternative='two-sided');

    print(tagname+":");
    fmtstr="pvals:\n all:\t{:1.2f}(exp:{:3.1f}%,obs:{:3.1f}%)\n cases:\t{:1.2f}(exp:{:3.1f}%,obs:{:3.1f}%)\n controls:\t{:1.2f}(exp:{:3.1f}%,obs:{:3.1f}%)";
    print("census={}%,our population={:3.1f}%".format(census*100, 100*pval));
    print(fmtstr.format(allSig, (census*100), (100*popTotalCurrent/popTotalPatients),
                        casesSig, (pval*100), (100*casesTotalCurrent/AETotalPatients),
                        controlsSig, (pval*100), (100*controlsTotalCurrent/AETotalPatients)  ));


def welch_ttest(x, y): 
    ## Welch-Satterthwaite Degrees of Freedom ##
    dof = (x.var()/x.size + y.var()/y.size)**2 / ((x.var()/x.size)**2 / (x.size-1) + (y.var()/y.size)**2 / (y.size-1))
    
    t, p = stats.ttest_ind(x, y, equal_var = False)
    
    print("\n",
          f"Welch's t-test= {t:.4f}", "\n",
          f"p-value = {p:.4f}", "\n",
          f"Welch-Satterthwaite Degrees of Freedom= {dof:.4f}")
    


patvarFilename="../out/DOAC/all/bleeding.patvars"
df = pd.read_csv(patvarFilename,
                 sep="\t", 
                 names=["pid", "age", "yearsDeceased", "sex", "caucasion", "ethnic", "married", "religion", "zip", "zippath", "AEFlag"]);

popTotalPatients=int(df.describe().loc['count','age']);

#
# convert from categorical strongs to binary integers:
#
df.loc[df['sex']=='M','sex']=0;
df.loc[df['sex']=='F','sex']=1;

df.loc[df['caucasion']!='1','caucasion']=0;
df.loc[df['caucasion']!=0,'caucasion']=1;

print(df.head());
aestats=df[df["AEFlag"]==2].describe();
AETotalPatients=int(aestats.loc["count","pid"]);
print("Total adverse events [4338] = " + str(AETotalPatients) );

cases = df[(df['AEFlag'] == 2)]
#cases=df[df["AEFlag"]==2]
print("*********CASES",end="");
numCases=cases.describe().loc['count','age']
print("["+str(numCases)+"]");
print(cases.describe().loc[['mean','min','max'],'age':'yearsDeceased']);
print("Age: (W test statistic, p-value):" + str(stats.shapiro(cases['age'])) );
print("Years Deceased: (W test statistic, p-value):" + str(stats.shapiro(cases['yearsDeceased'])) );

controls=df[(df["AEFlag"]==1)].sample(n=AETotalPatients, random_state=2)
print("*****CONTROLS",end="");
numControls=controls.describe().loc['count','age']
print("["+str(numControls)+"]");
print(controls.describe().loc[['mean','min','max'],'age':'yearsDeceased']);
print("Age: (W test statistic, p-value):" + str(stats.shapiro(controls['age'])) );
print("Years Deceased: (W test statistic, p-value):" + str(stats.shapiro(controls['yearsDeceased'])) );

print();
print("My T-test on age:",end="");
welch_ttest(cases['age'], controls['age']);

print("T-test on age:" + str(stats.ttest_ind(cases['age'], controls['age'], equal_var = False)));

print();


print("---Binomial tests, low p-val means we reject the hypothesis that the observed mean is within the expected range): ");
print("   In other words, low p-val means the population isn't balanced");
print("The original value is {:d}% controls, so the pval for controls should be low.".format(int(100*(popTotalPatients-AETotalPatients)/popTotalPatients)));

print();
compare_dist(df=df, cases=cases, controls=controls, census=.508, colname="sex", tagname="FEMALE");
print();
compare_dist(df=df, cases=cases, controls=controls, census=.72, colname="caucasion", tagname="CAUCASION");
print();
compare_dist(df=df, cases=cases, controls=controls, census=.34, colname="religion", tagname="RELIGION");

print(cases.groupby(["religion"]).size().sort_values().tail(10));
print(df.groupby(["religion"]).size().sort_values().tail(10));


procCSV/xlate-ae-patvars.pl
#!/usr/bin/perl
use warnings;
use strict;
use 5.010;

use CMDUtil::lib::CMDUtil;

##################
# Set-up
##################
# read-in the default values from config.json
my $settingsFilename = './procCSV/config.json';
# the json has config for multiple programs, just read params for "filter-pids"
our $config = "xlate-ae-patvars";

our %opt;
our $dimFilename;
# now read the config file and get the defaults

CMDUtil::getDefaults($settingsFilename, $config);
%opt=%CMDUtil::opt;


$dimFilename = $opt{'baseDim'}.$opt{'drugType'}.$opt{'fileDim'};

use Getopt::Long qw< :config auto_version bundling no_ignore_case >;
use Pod::Usage qw(pod2usage);


# get any override values from command line, and/or print documentation
GetOptions ("e|encoding=s"   => \$opt{'encoding'},
	    "a|aeType=s"     => \$opt{'aeType'},
	    "d|drugType=s"   => \$opt{'drugType'},
	    "o|outpath=s"    => \$opt{'outPath'},
	    "i|inputFile=s"  => \$opt{'dimFilename'},
	    "b|baseDim=s"    => \$opt{'baseDim'},
	    "f|fileDim=s"    => \$opt{'fileDim'},
	    "s|settingsFile=s" => \$opt{'settingsFilename'},
	    "y|year=i"       => \$opt{'currentYear'},
	    "r|dryRun!"      => \$opt{'dryrun'},
            "v|verbose"      => \$opt{'verbose'}, 

    # Standard options
	    "usage"        => sub { pod2usage(2) },
	    "help|?"       => sub { pod2usage(1) },
	    "manual"       => sub {pod2usage( -exitstatus => 0, -verbose => 2) },
    )
    or pod2usage(2);
pod2usage(1) if $opt{'help'};
pod2usage(-exitval => 0, -verbose => 2) if $opt{'man'};


# re-read configuration if user specified a differieng json file
if(defined $opt{'settingsFilename'} && $opt{'settingsFilename'} ne "") {
    CMDUtil::info("Reading new defaults from:$opt{'settingsFilename'}");
    getDefaults($opt{'settingsFilename'});
} else {
    $opt{'settingsFilename'} = $settingsFilename;
}

# if the input file isn't explicitly specified, construct it from other options
if( ! defined $opt{'dimFilename'} || $opt{'dimFilename'} eq "") {
    $opt{'dimFilename'}="$opt{'baseDim'}$opt{'drugType'}/$opt{'fileDim'}";
    CMDUtil::info("dimFilename is NOT defined, or else it is defined but it's empty, set using baseDim, drugType, fileDim options: $opt{'dimFilename'}");
}

# Finally, dump out the configuration
%CMDUtil::opt = %opt;
CMDUtil::info("--- SETTINGS: ---"); # only prints if verbose is actually on
use Data::Dumper;
CMDUtil::info( Dumper \%opt) ;

# start the dry run
if( $opt{'dryrun'} ) {
    CMDUtil::info("\n******************************* DRY RUN **********************************");
}

##################
# Parse
##################

my $pidsPath="$opt{'outPath'}/$opt{'drugType'}/$opt{'encoding'}";

# test that the pids files exist and are readable
( -e $pidsPath && -d _ ) or  die("Can't write files to $pidsPath");
my $matchName="${pidsPath}/$opt{'aeType'}.pid-code";
my $notMatchName="${pidsPath}/NOT$opt{'aeType'}.pid-code";
( -e $matchName) or die ("$matchName doesn't exist");

# if you're translating an ICD code from filter-pids, you'll have a 'NOT<aeType>' file, 
# if if you're translating a file that's a concat of all the ae's from ICD9, ICD10, you won't have a 'NOT<aetype> fil.
my $existsNOTeaType = 0;
if( -e $notMatchName){ $existsNOTeaType = 1; } else {
    CMDUtil::info("No file for NOT<aeType>, just use ae pids as a filter without any checking.");
}

my $count;

my @aePids; 

CMDUtil::info("get pids");

chomp( @aePids =`cut -f 1 ${matchName}|sort -u`); die "awk/sort failed: $?" if $?;
CMDUtil::info("got ".($#aePids+1)." pids (with adverse events): ${matchName}\n");

my @NOTaePids;

if($existsNOTeaType) {

    CMDUtil::info("+ cut -f 1 ${notMatchName}|sort -u");
    $count = `wc -l < ${notMatchName}`; die "wc failed: $?" if $?; chomp($count);
    CMDUtil::info("line count = $count, this might take a minute, please wait...");

    my @unfilteredNOTaePids;

    chomp( @unfilteredNOTaePids =`cut -f 1 ${notMatchName}|sort -u`); die "awk/sort failed: $?" if $?;
    CMDUtil::info("got ".($#unfilteredNOTaePids+1)." pids with non-adverse events ${notMatchName}\n");
    my %hNOTaePids = map { my $x = $_; $x => (! grep(/^${x}$/, @aePids) ? 1 : 0) }  @unfilteredNOTaePids; 
    @NOTaePids = grep { $hNOTaePids{$_} eq '1' } keys %hNOTaePids;
    CMDUtil::info("got ".($#NOTaePids+1)." pids that never had an adverse event ${notMatchName}\n");
}

# these constants could maybe be in the config:
use constant MAX_LINES => 300000000; # ~ 70GB, for the STDIN case where you can't compute the lines in advance
use constant STEP_SIZE => 1000; # number of lines to read before updating the progress bar
my $fh;
if($opt{'dimFilename'} eq "-") {
    $fh = *STDIN;
    $count = MAX_LINES;
    CMDUtil::info("reading from STDIN");
} else {
    open($fh,"<",$opt{'dimFilename'}) or die ("Can't open input file $opt{'dimFilename'}");
    $count = `wc -l < $opt{'dimFilename'}`; die "wc failed: $?" if $?; chomp($count);
    CMDUtil::info("reading from $opt{'dimFilename'}");
}

my $currentYear = 2019;
my $header=<$fh>;
#print "pid\tage\tyearsDeceased\tsex\traceCode\tethnicityCode\tmaritalStatusCode\treligionCode\tzipCode\tstateCityZip_path\n";

use Term::ProgressBar;
my $progress = Term::ProgressBar->new($count);
my $lineNum=0;
use constant CONTROL => 1;
use constant ADVERSE_EVENT => 2;
while( <$fh>) {
    $lineNum++;
    if( ($lineNum % STEP_SIZE) == 0) {
        $progress->update($lineNum);
    }

    if( ! $opt{'dryrun'} ){

	my ($pid, $vitalsCode, $birthDate, $deathDate, $sex, $languageCode, $raceCode, $ethnicityCode, $maritalStatusCode, $religionCode, 
	    $zipCode, $stateCityZip_path, $primeInsClsCd, $primeInsPayerCode, $steDeatDate, $steDeatWms, $updateDate, 
	    $downloadDate, $importDate, $sourceSystmeCode, $uploadId, $deathDt) = $_ =~ m/("[^"]+"|[^,]+)(?:,\s*)?/g;
	
	my $AEFlag=0;
	if($pid ~~ @aePids) { $AEFlag=ADVERSE_EVENT; }
	if($existsNOTeaType) {
	    if($pid ~~ @NOTaePids) { 
		if($AEFlag == ADVERSE_EVENT) { die("Error: patient was found in adverse event AND controls list: \npid=$pid\n$matchName\n$notMatchName");}
		$AEFlag=CONTROL; 
	    }
	} else {
	    if(!$AEFlag) {$AEFlag = CONTROL;}
	}

	$birthDate =~ s/"//g;
	$deathDate =~ s/"//g;
	$sex =~ s/"//g;
	
	my $yearsDeceased=0;
	if($deathDate ne "(null)") {
	    my($deathDay, @scrap)=split(" ", $deathDate);
	    my($year, $month, $day)=split("/", $deathDay);
	    $yearsDeceased = $opt{"currentYear"} - $year;
	}
	my($birthDay, @scrap)=split(" ", $birthDate);
	my ($year, $month, $day)=split("/", $birthDay);
	my $age = $opt{"currentYear"} - $year - $yearsDeceased;
	
	if($AEFlag){
	    print "$pid\t$age\t$yearsDeceased\t$sex\t$raceCode\t$ethnicityCode\t$maritalStatusCode\t$religionCode\t$zipCode\t$stateCityZip_path\t$AEFlag\n";
	} else {
	    CMDUtil::info(" ! $pid not in adverse or controls");
	    print "$pid\t$age\t$yearsDeceased\t$sex\t$raceCode\t$ethnicityCode\t$maritalStatusCode\t$religionCode\t$zipCode\t$stateCityZip_path\t".ADVERSE_EVENT."\n";
	}
    }

}

$progress->update($count);
CMDUtil::info("-----Done. Next: analysis -----");


######################### Documentation ######################### 

__END__

=head1 NAME

  xlate-ae-patvars.pl - creates table of clinical patient variables tagged with whether they encountered an adverse event (1) or not (2) during the study period.

=head1 SYNOPSIS

 ./xlate-ae-patvars.pl [--encoding <ICD10|ICD9>] [--aeType <ae-type>] [--drugType <DOAC|GENT>] [--inputFile <dx-filename>] [--help] [--verbose] 

=head1 DESCRIPTION

B<xlate-ae-patvars.pl>  data will be output to STDOUT

The following is an example pipeline: 

#1.
    ./procCSV/filter-pids.pl -e ICD10 -a bleeding -d DOAC
#2.
    head -100 /data/CSV/GONZ_<drugType>/patient_dimension.csv| ./procCSV/xlate-ae-patvars.pl  -e ICD10 -a bleeding -d DOAC -i - > ../out/DOAC/ICD10/bleeding.patvars

 reads 'ae' pids from <outroot>/<drugType>/<encoding>/<aeType>.pid
 reads 'NOTae' pids from <outroot>/<drugType>/<encoding>/NOT<aeType>.pid
 1 = AE
 2 = control


=head1 OPTIONS

=for pod2usage:
  help message for pod2usage

=over

=item B<-e>, B<--encoding>

Discerns which file to use in data.in/codes/<ae-type>.<encoding>.codes
                      [$opt{'encoding'}]    

=item B<-a>, B<--aeType>

Discerns which file to use in data.in/codes/<ae-type>.<encoding>.codes
                      [$opt{'aeType'}] 

=item B<-d>, B<--drugType>

Discerns which drug type to use (e.g., DOAC or GENT)
                      [$opt{'drugType'}]

=item B<-i>, B<--inputFile>

If '-', reads from STDIN. Overrides -b, -f (below)
                      [${dxFilename}]

=item B<-b>, B<--baseDim>

Base path in which to find the i2b2-sourced 'Patient dimension' CSV file
                      [$opt{'baseDim'}]

=item B<-f>, B<--fileDim>

Filename of the i2b2-sourced 'Patient dimension'  CSV file
                      [$opt{'fileDim'}]

=item B<-s>, B<--settingsFile>

Path to a json containing the programs default values for all the arguments described here. This file is found in B<./procCSV/config.json> by default.


=item B<-y>, B<--year>

Year relative to which the patient age will be calculated.


=item B<-r>, B<--dryrun>

Don't do any heavy lifting, just open files, create paths - don't read the big files, don't over-write anything, just print the 'informative' messages if verbose is on. Messages are prepended with '[D]' in dryrun mode.

=item B<-v>, B<--verbose>

Enable verbose mode

=item B<--man>

Print man page.

=back

=head1 TODO

Add in support for more different drugs
Add ICD cods for other adverse events
Make i2b2 file format configurable

=head1 AUTHOR

Kimberly Robasky E<lt>krobasky@gmail.com<gt>


=head1 BUGS

None known.

=head1 SEE ALSO

xlate-ae-patvars.pl

=head1 COPYRIGHT and LICENSE

MIT License
Copyright (c) 2019 Kimberly Robasky/RENCI, UNC Chapel Hill


=cut
